### DISCUSSION TITLE
Outlook

### DISCUSSION THREAD

**Discussion by SilverChinchilla** [cmu-sei/GHOSTS#128] [Q&A] [answered]
Howdy again! Having some issues seting up Outlook, have gotten it to open users email just fine but it won't generate emails. (I can send emails manually just fine). The OU is a little lengthy but looks accurate in the application JSON. App.Log file showing the following errors for outlook:
Incorrect number of email config array items - got 7, expected 8
at Infastructure\Email\EmailConfiguration.cs line 63
at Handlers\Outlook.cs line 99 

Don't want to start changing things without knowing where to start looking. Any help is always appreciated!

**Comment by SilverChinchilla** [comment]
Additional info,
Exchange server 2019
Microsoft Office 2019
downloaded and installed Redemption 6.2, placed dlls in the /lib directory on client ghosts folder

**Comment by sei-dupdyke** [comment]
What does your timeline config look like for this handler? This line `Incorrect number of email config array items - got 7, expected 8` indicates something missing from the array [processed here](https://github.com/cmu-sei/GHOSTS/blob/116bd467c85392a530278ab905370344119eb53f/src/Ghosts.Client/Infrastructure/Email/EmailConfiguration.cs#L77).

Command args for this should be:

0 = From address
1 = To
2 = Cc
3 = Bcc
4 = Subject
5 = Body
6 = Body Type (HTML|RTF|plaintext)            
7 = Attachment path

If you can, post up your handler here please?

**Reply by SilverChinchilla** [reply]
would it also have something to do with the "MachineId" in the instance folder being all "0"s?

**Reply by SilverChinchilla** [reply]
{
            "HandlerType": "Outlook",
            "Initial": "",
            "UtcTimeOn": "00:00:00",
            "UtcTimeOff": "24:00:00",
            "Loop": "True",
            "TimeLineEvents": [
                {
                    "Command": "create",
                    "CommandArgs": [
                        "CurrentUser",
                        "Random|Other:string ToEmailAddress - comma separate multiples",
                        "Random|Other:string CcEmailAddress - comma separate multiples",
                        "Random|Other:string BccEmailAddress - comma separate multiples",
                        "Random|Other:string Subject",
                        "Random|Other:string Body",
                        "PlainText|RTF|HTML enum BodyType",
                        "string Attachments - comma separate multiples"
                    ],
                    "DelayAfter": 900000,
                    "DelayBefore": 0
                },
                {
                    "Command": "reply",
                    "CommandArgs": [
                        "CurrentUser",
                        "All - reply to all",
                        "All",
                        "All",
                        "Parent - format is 'RE: <original message/>'",
                        "Random+Parent - format is reply then original message",
                        "Parent - format of original message",
                        ""
                    ],
                    "DelayAfter": 900000,
                    "DelayBefore": 0
                }
            ]
        },
        {
            "HandlerType": "Outlook",
            "Initial": "",
            "UtcTimeOn": "00:00:00",
            "UtcTimeOff": "24:00:00",
            "Loop": "True",
            "TimeLineEvents": [
                {
                    "Command": "create",
                    "CommandArgs": [
                        "CurrentUser",
                        "random",
                        "random",
                        "random",
                        "Random",
                        "Random",
                        "PlainText",
                        ""
                    ],
                    "DelayAfter": 900000,
                    "DelayBefore": 0
                },
                {
                    "Command": "reply",
                    "CommandArgs": [
                        "CurrentUser",
                        "All",
                        "All",
                        "All",
                        "Parent",
                        "Parent+Random",
                        "Parent",
                        ""
                    ],
                    "DelayAfter": 900000,
                    "DelayBefore": 0
                }

**Reply by SilverChinchilla** [reply]
so it opens up outlook but it doesn't create an email, wherever that is on the server that controls that I have no clue lol, but that is the point where I am at.

**Comment by sei-dupdyke** [comment]
> would it also have something to do with the "MachineId" in the instance folder being all "0"s?

Hmm, I have never seen this happen before. It will not affect email function though. Do you mind deleting this file and seeing if it gets correctly re-saved with an actual guid please?

The sample timeline has my poor attempts at inline documentation in there - the first section is likely the problem. Trim down your timeline to just having the following:

```
{
"HandlerType": "Outlook",
"Initial": "",
"UtcTimeOn": "00:00:00",
"UtcTimeOff": "24:00:00",
"Loop": "True",
"TimeLineEvents": [
{
"Command": "create",
"CommandArgs": [
"CurrentUser",
"random",
"random",
"random",
"Random",
"Random",
"PlainText",
""
],
"DelayAfter": 900000,
"DelayBefore": 0
},
{
"Command": "reply",
"CommandArgs": [
"CurrentUser",
"All",
"All",
"All",
"Parent",
"Parent+Random",
"Parent",
""
],
"DelayAfter": 900000,
"DelayBefore": 0
}
```

This should fix the issues.

**Reply by SilverChinchilla** [reply]
deleted the instance folder and logs folder and it did not create a GUID, going to trim the timeline and see if it helps

**Comment by sei-dupdyke** [comment, marked as answer]
In your application.json, is `"IdEnabled": true,`?

**Reply by SilverChinchilla** [reply]
it is enabled to True. And no luck on emails after trimming it to the above

**Reply by SilverChinchilla** [reply]
The EULA for Redemption keeps popping up to click "agree" would this cause a laps on the timeline where outlook won't execute properly?

**Reply by sei-dupdyke** [reply]
Oh, it could be redemption I suppose, (although I'd expect logs to show that more clearly) - try this:

In the lib folder, there is an installer.exe for redemption, step through that with all the defaults.

**Reply by SilverChinchilla** [reply]
the API is assigning the host an ID but just not saving it on the host files

**Reply by sei-dupdyke** [reply]
Is there anything in app.log regarding the ID?

**Reply by SilverChinchilla** [reply]
did the default install for Redemption but still pops up after executing ghosts. in the app.log it says that the ...CheckId.get_Id|config file could not be opened

**Reply by SilverChinchilla** [reply]
correction, not getting array error, but getting a PSDataCollection error along with the Id error

**Reply by sei-dupdyke** [reply]
What version of Windows?

**Reply by SilverChinchilla** [reply]
windows 10 (1607)

**Reply by sei-dupdyke** [reply]
🤔 truly stumped

1) id file not updating
2) outlook/redemption issues

**Reply by sei-dupdyke** [reply]
tested both of these things this morning repeatedly with no issues on 

Edition	Windows 10 Pro
Version	21H2
OS build	19044.2006
Experience	Windows Feature Experience Pack 120.2212.4180.0

**Reply by SilverChinchilla** [reply]
i did a buildout test range on 21h2 a couple months ago and got it working but wanted to try it on older version of Windows and can't seem to figure it out

**Reply by sei-dupdyke** [reply]
Do other handlers—say browsing—work on this version of windows?

**Reply by SilverChinchilla** [reply]
yes sir, every other app (browser, powerpoint, excel, word, commandline) all work perfect

**Reply by SilverChinchilla** [reply]
only thing i can think of trying next is downgrading the edition of redemption

**Reply by SilverChinchilla** [reply]
update: older version of redemption does not work either

**Reply by sei-dupdyke** [reply]
When you run the redemption installer, does it error at all?

**Reply by SilverChinchilla** [reply]
no errors, goes to default in Program Files(x86), no delay, nothing. EULA continues to pop up but no other errors showing

**Reply by SilverChinchilla** [reply]
I am gunna continue downgrading redemption, once I run through all of those then that at lest crosses off that variable

**Reply by SilverChinchilla** [reply]
update:still nothing

**Reply by SilverChinchilla** [reply]
for the OU in the application.json, if on Active directory there are users in 3 domains, but it is just actually one flat domain, would that effect how i list the OU structure? Or does that not even matter for this case?

**Reply by sei-dupdyke** [reply]
Without ghosts doing any automation, can you send and receive email from Outlook manually?

**Reply by SilverChinchilla** [reply]
yes, without ghosts, outlook works perfectly fine

**Reply by SilverChinchilla** [reply]
would any of this be solved with a fresh Ubuntu Install? or is it 99% client side

**Reply by SilverChinchilla** [reply]
new error is popping up for the outlook portion of timeline, ExecuteEvents...Unexpected Character encountered while parsing value 1. Path [36], line 38 position 12

**Reply by sei-dupdyke** [reply]
Do you mind sending me your app.log file(s)? If you want, just send them to my email, pls. Thanks.
